File: src/mdblocks.js
/**
 * src/mdblocks.js — a small Markdown → InputRichBlock converter.
 *
 * Used when a post must be sent in *blocks* mode (e.g. slideshows built from
 * an uploaded album, where every photo is embedded by file_id inside an
 * InputRichBlockPhoto instead of a `tg://photo?id=` reference).
 *
 * Supported (a pragmatic subset of Rich Markdown):
 *   # .. ###### headings           > blockquote lines
 *   - / * bullet lists             1. ordered lists
 *   ```` ``` fenced code ``` ````  --- divider
 *   **bold** *italic* `code` ==marked== ||spoiler|| [text](url) ![alt](url)
 * Anything else becomes a paragraph.
 */

import * as R from './rich.js';

const INLINE = [
  ['code', /`([^`\n]+)`/g],
  ['bold', /\*\*([^*\n]+)\*\*/g],
  ['italic', /(?<!\*)\*([^*\n]+)\*(?!\*)/g],
  ['marked', /==([^=\n]+)==/g],
  ['spoiler', /\|\|([^|\n]+)\|\|/g],
  ['url', /\[([^\]\n]+)\]\((https?:\/\/[^)\s]+)\)/g],
];

const IMG = /!\[[^\]]*\]\(([^)\s]+)\)/g;

/** Convert one line of inline markdown into a RichText value. */
export function inline(text) {
  let s = String(text ?? '').replace(IMG, '');
  if (!s) return '';
  const tokens = [];
  for (const [kind, pattern] of INLINE) {
    pattern.lastIndex = 0;
    let m;
    while ((m = pattern.exec(s)) !== null) {
      tokens.push({ start: m.index, end: m.index + m[0].length, kind, groups: m.slice(1) });
    }
  }
  tokens.sort((a, b) => a.start - b.start);
  const out = [];
  let last = 0;
  for (const tok of tokens) {
    if (tok.start < last) continue;           // overlapping — skip
    if (tok.start > last) out.push(s.slice(last, tok.start));
    const [g1, g2] = tok.groups;
    if (tok.kind === 'code') out.push(R.code(g1));
    else if (tok.kind === 'bold') out.push(R.bold(g1));
    else if (tok.kind === 'italic') out.push(R.italic(g1));
    else if (tok.kind === 'marked') out.push(R.marked(g1));
    else if (tok.kind === 'spoiler') out.push(R.spoiler(g1));
    else if (tok.kind === 'url') out.push(R.link(g1, g2));
    last = tok.end;
  }
  if (last < s.length) out.push(s.slice(last));
  if (!out.length) return '';
  if (out.length === 1 && typeof out[0] === 'string') return out[0];
  return out;
}

/** Convert Rich Markdown text into an array of InputRichBlock dicts. */
export function mdToBlocks(markdown) {
  const blocks = [];
  const lines = String(markdown ?? '').replace(/\r\n/g, '\n').split('\n');
  let i = 0;
  let paragraph = [];

  const flushParagraph = () => {
    if (paragraph.length) {
      const text = paragraph.filter(Boolean).join(' ');
      if (text.trim()) blocks.push(R.paragraph(inline(text)));
      paragraph = [];
    }
  };

  while (i < lines.length) {
    const line = lines[i];
    const stripped = line.trim();

    if (!stripped) { flushParagraph(); i += 1; continue; }

    if (stripped.startsWith('```')) {
      flushParagraph();
      const language = stripped.slice(3).trim() || null;
      const codeLines = [];
      i += 1;
      while (i < lines.length && !lines[i].trim().startsWith('```')) { codeLines.push(lines[i]); i += 1; }
      i += 1; // closing fence
      blocks.push(R.pre(codeLines.join('\n'), language));
      continue;
    }

    let m = stripped.match(/^(#{1,6})\s+(.*)$/);
    if (m) { flushParagraph(); blocks.push(R.heading(inline(m[2]), m[1].length)); i += 1; continue; }

    if (/^(-{3,}|\*{3,}|_{3,})$/.test(stripped)) { flushParagraph(); blocks.push(R.divider()); i += 1; continue; }

    if (stripped.startsWith('>')) {
      flushParagraph();
      const quote = [];
      while (i < lines.length && lines[i].trim().startsWith('>')) {
        quote.push(lines[i].trim().replace(/^>/, '').trim());
        i += 1;
      }
      const body = quote.filter(Boolean).join(' ');
      blocks.push(R.blockquote(body ? [R.paragraph(inline(body))] : []));
      continue;
    }

    m = stripped.match(/^[-*+]\s+(.*)$/);
    if (m) {
      flushParagraph();
      const items = [];
      while (i < lines.length) {
        const mm = lines[i].trim().match(/^[-*+]\s+(.*)$/);
        if (!mm) break;
        let itemText = mm[1];
        let checked = null;
        const cm = itemText.match(/^\[([ xX])\]\s+(.*)$/);
        if (cm) { checked = cm[1].toLowerCase() === 'x'; itemText = cm[2]; }
        items.push(R.listItem(inline(itemText), { checkbox: checked !== null, checked: !!checked }));
        i += 1;
      }
      blocks.push(R.bulletList(items));
      continue;
    }

    m = stripped.match(/^(\d+)[.)]\s+(.*)$/);
    if (m) {
      flushParagraph();
      const items = [];
      while (i < lines.length) {
        const mm = lines[i].trim().match(/^(\d+)[.)]\s+(.*)$/);
        if (!mm) break;
        items.push(R.listItem(inline(mm[2])));
        i += 1;
      }
      blocks.push(R.orderedList(items));
      continue;
    }

    // table?
    if (stripped.startsWith('|') && stripped.endsWith('|')
        && i + 1 < lines.length && /^\|?[\s:|-]+\|?$/.test(lines[i + 1].trim())) {
      flushParagraph();
      const splitRow = (row) => row.trim().replace(/^\||\|$/g, '').split('|').map((c) => c.trim());
      const header = splitRow(lines[i]);
      const sep = splitRow(lines[i + 1]);
      i += 2;
      const aligns = header.map((_, idx) => {
        const cellTxt = sep[idx] || '';
        if (cellTxt.startsWith(':') && cellTxt.endsWith(':')) return 'center';
        if (cellTxt.endsWith(':')) return 'right';
        return 'left';
      });
      const body = [header];
      while (i < lines.length && lines[i].trim().startsWith('|')) {
        body.push(splitRow(lines[i].trim()));
        i += 1;
      }
      blocks.push(R.table(body, { header: true, aligns }));
      continue;
    }

    paragraph.push(stripped);
    i += 1;
  }

  flushParagraph();
  return blocks;
}

export default mdToBlocks;

File: apb/mdblocks.py
"""apb.mdblocks — a small Markdown -> InputRichBlock converter.

Used when a post must be sent in *blocks* mode (e.g. slideshows built from
uploaded albums, where photos are embedded by file_id directly in
``InputRichBlockPhoto`` objects instead of ``tg://photo?id=`` references).

Supported (a pragmatic subset of Rich Markdown):
  # .. ###### headings        > blockquote lines
  - / * bullet lists          1. ordered lists
  ``` fenced code blocks ```  --- divider
  **bold** *italic* `code` ==marked== ||spoiler||
  [text](url) ![alt](url)
Everything else becomes a paragraph.
"""

from __future__ import annotations

import re

from . import rich as R

_INLINE = [
    ("code", re.compile(r"`([^`\n]+)`")),
    ("bold", re.compile(r"\*\*([^*\n]+)\*\*")),
    ("italic", re.compile(r"(?<!\*)\*([^*\n]+)\*(?!\*)")),
    ("marked", re.compile(r"==([^=\n]+)==")),
    ("spoiler", re.compile(r"\|\|([^|\n]+)\|\|")),
    ("url", re.compile(r"\[([^\]\n]+)\]\((https?://[^)\s]+)\)")),
]

_IMG = re.compile(r"!\[[^\]]*\]\(([^)\s]+)\)")


def _inline(text):
    """Convert one line of inline markdown into a RichText value."""
    # images inside text are dropped in blocks mode (media handled separately)
    for pat, repl in ((re.compile(r"!\[[^\]]*\]\([^)]*\)"), ""),
                      (re.compile(r"\[([^\]]+)\]\(([^)]+)\)"), r"\1")):
        pass  # links handled by _INLINE; images removed below
    text = _IMG.sub("", text)
    if not text:
        return ""
    out = []
    pos = 0
    tokens = []  # (start, end, kind, payload)
    for kind, pattern in _INLINE:
        for m in pattern.finditer(text):
            tokens.append((m.start(), m.end(), kind, m.groups()))
    tokens.sort(key=lambda t: t[0])
    last = 0
    for start, end, kind, groups in tokens:
        if start < last:      # overlapping — skip
            continue
        if start > last:
            out.append(text[last:start])
        if kind == "code":
            out.append(R.code(groups[0]))
        elif kind == "bold":
            out.append(R.bold(groups[0]))
        elif kind == "italic":
            out.append(R.italic(groups[0]))
        elif kind == "marked":
            out.append(R.marked(groups[0]))
        elif kind == "spoiler":
            out.append(R.spoiler(groups[0]))
        elif kind == "url":
            out.append(R.link(groups[0], groups[1]))
        last = end
    if last < len(text):
        out.append(text[last:])
    if not out:
        return ""
    if len(out) == 1 and isinstance(out[0], str):
        return out[0]
    return out


def md_to_blocks(markdown):
    """Convert Rich Markdown text into a list of InputRichBlock dicts."""
    blocks = []
    lines = (markdown or "").replace("\r\n", "\n").split("\n")
    i = 0
    paragraph = []

    def flush_paragraph():
        if paragraph:
            text = " ".join(p for p in paragraph if p)
            if text.strip():
                blocks.append(R.paragraph(_inline(text)))
            paragraph.clear()

    while i < len(lines):
        line = lines[i]
        stripped = line.strip()

        if not stripped:
            flush_paragraph()
            i += 1
            continue

        if stripped.startswith("```"):
            flush_paragraph()
            code_lines = []
            i += 1
            language = stripped[3:].strip() or None
            while i < len(lines) and not lines[i].strip().startswith("```"):
                code_lines.append(lines[i])
                i += 1
            i += 1  # closing fence
            blocks.append(R.pre("\n".join(code_lines), language))
            continue

        m = re.match(r"^(#{1,6})\s+(.*)$", stripped)
        if m:
            flush_paragraph()
            blocks.append(R.heading(_inline(m.group(2)), len(m.group(1))))
            i += 1
            continue

        if re.match(r"^(-{3,}|\*{3,}|_{3,})$", stripped):
            flush_paragraph()
            blocks.append(R.divider())
            i += 1
            continue

        if stripped.startswith(">"):
            flush_paragraph()
            quote = []
            while i < len(lines) and lines[i].strip().startswith(">"):
                quote.append(lines[i].strip().lstrip(">").strip())
                i += 1
            blocks.append(R.blockquote(
                [R.paragraph(_inline(" ".join(q for q in quote if q)))] if quote else []))
            continue

        m = re.match(r"^[-*+]\s+(.*)$", stripped)
        if m:
            flush_paragraph()
            items = []
            while i < len(lines):
                mm = re.match(r"^[-*+]\s+(.*)$", lines[i].strip())
                if not mm:
                    break
                item_text = mm.group(1)
                checked = None
                cm = re.match(r"^\[([ xX])\]\s+(.*)$", item_text)
                if cm:
                    checked = cm.group(1).lower() == "x"
                    item_text = cm.group(2)
                items.append(R.list_item(_inline(item_text), checkbox=checked is not None,
                                         checked=bool(checked)))
                i += 1
            blocks.append(R.bullet_list(items))
            continue

        m = re.match(r"^(\d+)[.)]\s+(.*)$", stripped)
        if m:
            flush_paragraph()
            items = []
            while i < len(lines):
                mm = re.match(r"^(\d+)[.)]\s+(.*)$", lines[i].strip())
                if not mm:
                    break
                items.append(R.list_item(_inline(mm.group(2))))
                i += 1
            blocks.append(R.ordered_list(items))
            continue

        # table row?
        if stripped.startswith("|") and stripped.endswith("|") and i + 1 < len(lines) \
                and re.match(r"^\|?[\s:|-]+\|?$", lines[i + 1].strip()):
            flush_paragraph()
            rows = []

            def split_row(row):
                cells = [c.strip() for c in row.strip().strip("|").split("|")]
                return cells

            header = split_row(lines[i])
            sep = split_row(lines[i + 1].strip())
            i += 2  # skip separator
            aligns = []
            for idx in range(len(header)):
                cell = sep[idx] if idx < len(sep) else ""
                if cell.startswith(":") and cell.endswith(":"):
                    aligns.append("center")
                elif cell.endswith(":"):
                    aligns.append("right")
                else:
                    aligns.append("left")
            body = [header]
            while i < len(lines) and lines[i].strip().startswith("|"):
                body.append(split_row(lines[i].strip()))
                i += 1
            blocks.append(R.table(body, header=True, aligns=aligns))
            continue

        paragraph.append(stripped)
        i += 1

    flush_paragraph()
    return blocks
